pkg/provider/oidc.go

```diff
@@ -0,0 +1,195 @@
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/zitadel/oidc/v3/pkg/client/rp"
+	"github.com/zitadel/oidc/v3/pkg/client/rs"
+	httphelper "github.com/zitadel/oidc/v3/pkg/http"
+	"github.com/zitadel/oidc/v3/pkg/oidc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/zitadel/zitadel-go/v2/pkg/info"
+)
+
+var (
+	ErrMissingHeader        = errors.New("auth header missing")
+	ErrInvalidHeader        = errors.New("invalid auth header")
+	ErrInvalidToken         = errors.New("invalid token")
+	ErrInvalidAuthorization = errors.New("invalid authorization")
+)
+
+func newRelayingPartyOIDC(ctx context.Context, oidc *configurationOIDC) (rp.RelyingParty, error) {
+	redirectURI := fmt.Sprintf("http://localhost:%v%v", oidc.port, oidc.callbackURL)
+	cookieHandler := httphelper.NewCookieHandler(key, key, httphelper.WithUnsecure())
+	options := []rp.Option{
+		rp.WithCookieHandler(cookieHandler),
+		rp.WithVerifierOpts(rp.WithIssuedAtOffset(5 * time.Second)),
+	}
+	if oidc.clientSecret == "" {
+		options = append(options, rp.WithPKCE(cookieHandler))
+	}
+	if oidc.keyPath != "" {
+		options = append(options, rp.WithJWTProfile(rp.SignerFromKeyPath(oidc.keyPath)))
+	}
+	return rp.NewRelyingPartyOIDC(ctx, oidc.issuer, oidc.clientID, oidc.clientSecret, redirectURI, oidc.scopes, options...)
+}
+
+func marshalUserinfo() rp.CodeExchangeUserinfoCallback[*oidc.IDTokenClaims, *oidc.UserInfo] {
+	return func(w http.ResponseWriter, r *http.Request, tokens *oidc.Tokens[*oidc.IDTokenClaims], state string, rp rp.RelyingParty, info *oidc.UserInfo) {
+		data, err := json.Marshal(info)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Write(data)
+	}
+}
```

Comment:
what's the purpose of this?

Reply:
can be removed was part of the example and then the function to call can be provided as parameter